File: solutions/290-word-pattern/solution.py
class Solution:
    def word_pattern(self, pattern: str, s: str) -> bool:
        words = s.split()
        if len(pattern) != len(words):
            return False

        char_to_word: dict[str, str] = {}
        word_to_char: dict[str, str] = {}

        for ch, word in zip(pattern, words):
            if ch in char_to_word:
                if char_to_word[ch] != word:
                    return False
            else:
                if word in word_to_char:
                    return False
                char_to_word[ch] = word
                word_to_char[word] = ch

        return True


if __name__ == "__main__":
    tests = [
        ("abba", "dog cat cat dog", True),
        ("abba", "dog cat cat fish", False),
        ("aaaa", "dog cat cat dog", False),
        ("abba", "dog dog dog dog", False),
        ("a", "dog", True),
    ]
    sol = Solution()
    for pattern, s, expected in tests:
        result = sol.word_pattern(pattern, s)
        status = "PASS" if result == expected else "FAIL"
        print(f"{status} | pattern={pattern!r}, s={s!r} => {result} (expected {expected})")

File: solutions/290-word-pattern/solution.go
package main

import (
	"fmt"
	"strings"
)

func wordPattern(pattern string, s string) bool {
	words := strings.Fields(s)
	if len(pattern) != len(words) {
		return false
	}

	charToWord := make(map[byte]string)
	wordToChar := make(map[string]byte)

	for i := 0; i < len(pattern); i++ {
		ch := pattern[i]
		word := words[i]
		if mapped, ok := charToWord[ch]; ok {
			if mapped != word {
				return false
			}
		} else {
			if _, ok := wordToChar[word]; ok {
				return false
			}
			charToWord[ch] = word
			wordToChar[word] = ch
		}
	}
	return true
}

func main() {
	tests := []struct {
		pattern  string
		s        string
		expected bool
	}{
		{"abba", "dog cat cat dog", true},
		{"abba", "dog cat cat fish", false},
		{"aaaa", "dog cat cat dog", false},
		{"abba", "dog dog dog dog", false},
		{"a", "dog", true},
	}

	for _, tc := range tests {
		result := wordPattern(tc.pattern, tc.s)
		status := "PASS"
		if result != tc.expected {
			status = "FAIL"
		}
		fmt.Printf("%s | pattern=%q, s=%q => %v (expected %v)\n", status, tc.pattern, tc.s, result, tc.expected)
	}
}
